### PR TITLE
Add Jake Wharton's SDK location resolver code so it'll look in 

### PR DESCRIPTION
**Pull request by kenkyee**
local.properties for Android Studio

Luckily his sdk-manager is open source :-)

### CODE REVIEW

File: src/main/groovy/com/getkeepsafe/dexcount/DexFile.groovy [hunk 1]
```diff
@@ -80,16 +80,24 @@ class DexFile {
             }
         }
         // convert it to DEX format by using the Android dx tool
-        def androidSdkHome = System.getenv("ANDROID_HOME")
+        def androidSdkHome = DexMethodCountPlugin.sdkLocation
         if (androidSdkHome == null) {
             throw new Exception("ANDROID_HOME env variable not defined!")
```
Comment (xiphirx) [inline src/main/groovy/com/getkeepsafe/dexcount/DexFile.groovy]: This exception is no longer correct 100% of the time. This check would also be best suited in `DexMethodCountPlugin` when you first initialize the location.

File: src/main/groovy/com/getkeepsafe/dexcount/DexFile.groovy [hunk 2]
```diff
@@ -80,16 +80,24 @@ class DexFile {
...
         }
+
         def buildToolsSubDirs = new File(androidSdkHome, "build-tools")
         def dirs = buildToolsSubDirs.listFiles()
```
Comment (xiphirx) [inline src/main/groovy/com/getkeepsafe/dexcount/DexFile.groovy]: It seems like `def dirs = buildToolsSubDirs.listFiles().sort { it.name }.reverse()` would be better here. Removes the need for the for loop below.

File: src/main/groovy/com/getkeepsafe/dexcount/SdkResolver.groovy [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+import org.gradle.api.tasks.StopExecutionException
+
+import static com.android.SdkConstants.ANDROID_HOME_ENV
+import static com.android.SdkConstants.FN_LOCAL_PROPERTIES
+import static com.android.SdkConstants.PLATFORM_WINDOWS
+import static com.android.SdkConstants.SDK_DIR_PROPERTY
+import static com.android.SdkConstants.currentPlatform
+
+/**
+ * This was lifted and modified from Jake Wharton's SDK Manager gradle plugin
```
Comment (xiphirx) [inline src/main/groovy/com/getkeepsafe/dexcount/SdkResolver.groovy]: Lets have a link to the github repo this is from

File: src/main/groovy/com/getkeepsafe/dexcount/DexFile.groovy (hunk 1 above)
Comment (kenkyee) [inline src/main/groovy/com/getkeepsafe/dexcount/DexFile.groovy]: the SDK location is only needed if you process an AAR file...so that's why it's here instead...I'll update the msg

File: src/main/groovy/com/getkeepsafe/dexcount/SdkResolver.groovy [hunk 4]
```diff
@@ -0,0 +1,109 @@
+package com.getkeepsafe.dexcount
```
Comment (xiphirx) [inline src/main/groovy/com/getkeepsafe/dexcount/SdkResolver.groovy]: Let's put Jake's license header here and in `System`

**Comment by xiphirx** [comment]
Also a squash to one commit would be cool, but not necessary. Thanks!

File: src/main/groovy/com/getkeepsafe/dexcount/SdkResolver.groovy (hunk 4 above)
Comment (kenkyee) [inline src/main/groovy/com/getkeepsafe/dexcount/SdkResolver.groovy]: added

File: src/main/groovy/com/getkeepsafe/dexcount/DexMethodCountTask.groovy [hunk 5]
```diff
@@ -129,7 +129,7 @@ class DexMethodCountTask extends DefaultTask {
         printToFile(outputFile) { PrintStream out ->
             print(tree, out)
         }
-        outputTime = System.currentTimeMillis()
+        outputTime = java.lang.System.currentTimeMillis()
```
Comment (benjamin-bader) [inline src/main/groovy/com/getkeepsafe/dexcount/DexMethodCountTask.groovy]: Why the change?  This seemed to be working properly in the past.

**Comment by kenkyee** [comment]
No idea how to squash to one commit.  I always thought that was a github setting?  My other closed PR has a crazy huge amount of commits to debug that weird Travis issue :-)

File: src/main/groovy/com/getkeepsafe/dexcount/System.groovy [hunk 6]
```diff
@@ -0,0 +1,38 @@
+package com.getkeepsafe.dexcount
```
Comment (benjamin-bader) [inline src/main/groovy/com/getkeepsafe/dexcount/System.groovy]: Let's keep the original package names, ideally the original file content as well.

On the latter point, what did we have to change here, besides the package name?

File: src/main/groovy/com/getkeepsafe/dexcount/SdkResolver.groovy (hunk 4 above)
Comment (benjamin-bader) [inline src/main/groovy/com/getkeepsafe/dexcount/SdkResolver.groovy]: Same comment as System.groovy

File: src/main/groovy/com/getkeepsafe/dexcount/DexMethodCountTask.groovy (hunk 5 above)
Comment (kenkyee) [inline src/main/groovy/com/getkeepsafe/dexcount/DexMethodCountTask.groovy]: Because of the inclusion of the System interface from the SDK Manager files.  Conflicts now because currentTimeMillis() isnt' in that interface.  I could have added that to that interface and made it a static on Real and changed this to System.Real.currentTimeMillis() but that didn't seem to buy anything.  Other alternative is I can rename the System file from the SDK Manager.

**Comment by xiphirx** [comment]
You can squash by `git rebase -i <last commit on master hash>` then following the on screen instructions

File: src/main/groovy/com/getkeepsafe/dexcount/DexMethodCountTask.groovy (hunk 5 above)
Comment (kenkyee) [inline src/main/groovy/com/getkeepsafe/dexcount/DexMethodCountTask.groovy]: Should I rename his System interface then?  I was puzzled why he did that too.

File: src/main/groovy/com/getkeepsafe/dexcount/System.groovy (hunk 6 above)
Comment (kenkyee) [inline src/main/groovy/com/getkeepsafe/dexcount/System.groovy]: pulled out some references to other classes only in SdkResolver.  E.g., the ability to pull down the Android SDK.  Totally out of scope for this plugin.
System.groovy is exactly the same.  I can also push the two files down a package too...that should avoid the System class conflict I think?

Comment (benjamin-bader) [inline src/main/groovy/com/getkeepsafe/dexcount/System.groovy]: It *should*, yes; as long as it isn't in the same package as the task or higher, it won't cause problems elsewhere.

**Comment by benjamin-bader** [comment]
I think Github made squashing a repo option within the past week or two.  Personally I prefer squashing giant PRs, but not to the point of insisting on it.

File: src/main/groovy/com/getkeepsafe/dexcount/System.groovy (hunk 6 above)
Comment (kenkyee) [inline src/main/groovy/com/getkeepsafe/dexcount/System.groovy]: yep, it did...pushed those files down a package...that explains why they were down a package in Wharton's repo too :-)

**Comment by kenkyee** [comment]
FWIW, I couldn't get the squash to work.  I did "git rebase -i 380260e" and it did a bunch of stuff but never let me push the changes up...

**Comment by xiphirx** [comment]
Just seeing the comment about github's ability to do it automatically, seems to make my PR obsolete :P

**Comment by kenkyee** [comment]
We had a long discussion here about whether or not to turn that github feature on too.  It works w/ some people's PRs because they have a crapload of tiny commits while they're working.  For others, it's less so because they committed in big chunks of work.  FWIW, we decided against it in case we had to use git bisect to find out where something horked the code ;-)

**Comment by benjamin-bader** [comment]
Interesting.  The point which caused your team to avoid squashes is the exact same point that people use to argue *for* squashes - namely, that breaking changes are easy to back out; clean commit histories are not intrinsic goods, they are useful.

Not that I'm trying to sway you in any particular direction - it is actually interesting!

**Comment by kenkyee** [comment]
Interesting...so you can use git bisect on squashed commits and it'll still bisect into the hidden commits?  The worry was that you could only bisect down to the PR level since it would look like one big commit to git bisect...

**Comment by benjamin-bader** [comment]
No, there isn't such a thing as hidden commits; bisect will identify the single commit left behind by squash.

The notion that squashed commits are easier to back out is that tiny commits, in isolation, leave the branch in an unstable/untested state - many small commits, together, make for a coherent feature.  It is difficult in practice to revert part of a feature, while still keeping the build stable.  It is a much easier proposition to revert a whole feature.

**Comment by kenkyee** [comment]
ahh...that makes sense.  For some reason, people here wanted to be able to recover as much of the PR as possible instead of being stuck to PR granularity.  I can see both sides though I don't have a preference :-)

File: src/main/groovy/com/getkeepsafe/dexcount/sdkresolver/SdkResolver.groovy [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    final File localProperties
+    final boolean isWindows
+
+    SdkResolver(Project project, System system, boolean isWindows) {
+        this.project = project
+        this.system = system
+        this.isWindows = isWindows
+
+        userHome = new File(system.property('user.home'))
+        userAndroid = new File(userHome, '.android-sdk')
```
Comment (benjamin-bader) [inline src/main/groovy/com/getkeepsafe/dexcount/sdkresolver/SdkResolver.groovy]: I notice that, on my system at least, there is no `.android-sdk` directory.  Does this exist for you?  What installed it?

Also, Android Studio stores its SDK elsewhere; can we account for that here?

Comment (kenkyee) [inline src/main/groovy/com/getkeepsafe/dexcount/sdkresolver/SdkResolver.groovy]: That's a possible location.
Android Studio will update the local.properties file and put the location of the sdk in sdk.dir.  That's why I needed to do all this stuff.  They put that local.properties thing in for OSX users probably.  On Windows/Linux, you usually have ANDROID_HOME set up properly.

File: src/main/groovy/com/getkeepsafe/dexcount/sdkresolver/SdkResolver.groovy [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+import org.gradle.api.logging.Logging
+import org.gradle.api.tasks.StopExecutionException
+
+import static com.android.SdkConstants.ANDROID_HOME_ENV
+import static com.android.SdkConstants.FN_LOCAL_PROPERTIES
+import static com.android.SdkConstants.PLATFORM_WINDOWS
+import static com.android.SdkConstants.SDK_DIR_PROPERTY
+import static com.android.SdkConstants.currentPlatform
+
+/**
```
Comment (benjamin-bader) [inline src/main/groovy/com/getkeepsafe/dexcount/sdkresolver/SdkResolver.groovy]: The standard practice here is to keep the license at the very top of the file and, if anything has changed (which is the case here), adding a new line like so:

```
* Copyright 2016 KeepSafe Software
* Copyright 2014 Jake Wharton
* etc
```
Please do so with this and System.groovy.

Comment (kenkyee) [inline src/main/groovy/com/getkeepsafe/dexcount/sdkresolver/SdkResolver.groovy]: Didn't know that was standard practice :)
Updated both.

File: src/main/groovy/com/getkeepsafe/dexcount/sdkresolver/SdkResolver.groovy (hunk 7 above)
Comment (benjamin-bader) [inline src/main/groovy/com/getkeepsafe/dexcount/sdkresolver/SdkResolver.groovy]: I see, excellent.  I posted the comment b/c I noticed that I had `.android`, and not `.android-sdk`, and forgot what put that directory there!  `local.properties` handling is awesome.

**Comment by benjamin-bader** [comment]
OK, looks good - thanks again for the awesome effort!
